False positive map_clone on .map(|x| *x) on Option<Box<T>>

```rust
fn main() {
    let x = Some(Box::new(true));
    let y = x.map(|x| *x);
    println!("{:?}", y);
}
```

```console
warning: You are using an explicit closure for cloning elements                
 --> src/main.rs:3:13                                                          
  |                                                                            
3 |     let y = x.map(|x| *x);                                                 
  |             ^^^^^^^^^^^^^ help: Consider calling the dedicated `cloned` method: `x.cloned()`
  |                                                                            
  = note: #[warn(clippy::map_clone)] on by default                             
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#map_clone
```

The suggested code does not compile.

```console
error[E0599]: no method named `cloned` found for type `std::option::Option<std::boxed::Box<bool>>` in the current scope
 --> src/main.rs:3:15                                                          
  |                                                                            
3 |     let y = x.cloned();                                                    
  |               ^^^^^^                                                       
  |                                                                            
  = note: the method `cloned` exists but the following trait bounds were not satisfied:
          `&mut std::option::Option<std::boxed::Box<bool>> : std::iter::Iterator`
```

--- a/clippy_lints/src/map_clone.rs
+++ b/clippy_lints/src/map_clone.rs
@@ -79,7 +79,7 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Pass {
                         lint(cx, e.span, args[0].span, name, closure_expr);
                     },
                     hir::PatKind::Binding(hir::BindingAnnotation::Unannotated, _, name, None) => match closure_expr.node {
-                        hir::ExprKind::Unary(hir::UnOp::UnDeref, ref inner) => lint(cx, e.span, args[0].span, name, inner),
+                        hir::ExprKind::Unary(hir::UnOp::UnDeref, ref inner) if !cx.tables.expr_ty(inner).is_box() => lint(cx, e.span, args[0].span, name, inner),
                         hir::ExprKind::MethodCall(ref method, _, ref obj) => if method.ident.as_str() == "clone" && match_trait_method(cx, closure_expr, &paths::CLONE_TRAIT) {
                             lint(cx, e.span, args[0].span, name, &obj[0]);
                         }

--- a/tests/ui/map_clone.rs
+++ b/tests/ui/map_clone.rs
@@ -16,4 +16,5 @@ fn main() {
     let _: Vec<i8> = vec![5_i8; 6].iter().map(|x| *x).collect();
     let _: Vec<String> = vec![String::new()].iter().map(|x| x.clone()).collect();
     let _: Vec<u32> = vec![42, 43].iter().map(|&x| x).collect();
+    let _: Option<u64> = Some(Box::new(16)).map(|b| *b);
 }

